google-datacatalog-connectors-commons/src/google/datacatalog_connectors/commons/prepare/base_entry_factory.py

```diff
@@ -14,26 +14,48 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import hashlib
 import re
 import unicodedata
 import six
 
 
 class BaseEntryFactory:
     __ASCII_CHARACTER_ENCODING = 'ASCII'
+    __ID_MAX_LENGTH = 64
 
     @classmethod
     def _format_id(cls, source_id):
-        formatted_id = re.sub(r'[^a-zA-Z0-9]+', '_',
-                              cls.__normalize_ascii_chars(source_id.strip()))
-        return formatted_id[:64] if len(formatted_id) > 64 else formatted_id
+        formatted_id = cls.__format_and_normalize_str(r'[^a-zA-Z0-9]+',
+                                                      source_id)
+
+        return formatted_id[:cls.__ID_MAX_LENGTH] if \
+            len(formatted_id) > cls.__ID_MAX_LENGTH else formatted_id
+
+    @classmethod
+    def _format_id_with_hashing(cls, source_id, length_to_hash):
+        formatted_id = cls.__format_and_normalize_str(r'[^a-zA-Z0-9]+',
+                                                      source_id)
+
+        if len(formatted_id) <= cls.__ID_MAX_LENGTH:
+            return formatted_id
+
+        hash = hashlib.sha1()
+        hash.update(formatted_id.encode(cls.__ASCII_CHARACTER_ENCODING))
+
+        return formatted_id[:cls.__ID_MAX_LENGTH-length_to_hash] + \
+            hash.hexdigest()[:length_to_hash]
 
     @classmethod
     def _format_display_name(cls, source_name):
-        formatted_name = re.sub(
-            r'[^a-zA-Z0-9_\- ]+', '_',
-            cls.__normalize_ascii_chars(source_name).strip())
-        return formatted_name
+        return cls.__format_and_normalize_str(r'[^a-zA-Z0-9_\- ]+',
+                                              source_name)
+
+    @classmethod
+    def __format_and_normalize_str(cls, regex_pattern, source_str):
```

Comment:
```suggestion
    def __normalize_string(cls, regex_pattern, source_string):
```
I guess this name can be simplified with no context loss. WDYT? 🤔